RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitProducerKernel.dev.cc

```diff
@@ -0,0 +1,198 @@
+      return false;
+
+    return true;
+  }
+
+  template <>
+  ALPAKA_FN_ACC void PFRecHitProducerKernelConstruct<HCAL>::constructPFRecHit(
+      reco::PFRecHitDeviceCollection::View::element pfrh,
+      const HCAL::CaloRecHitSoATypeDevice::ConstView::const_element rh) {
+    pfrh.detId() = rh.detId();
+    pfrh.energy() = rh.energy();
+    pfrh.time() = rh.time();
+    pfrh.depth() = HCAL::getDepth(pfrh.detId());
+    const uint32_t subdet = getSubdet(pfrh.detId());
+    if (subdet == HcalBarrel)
+      pfrh.layer() = PFLayer::HCAL_BARREL1;
+    else if (subdet == HcalEndcap)
+      pfrh.layer() = PFLayer::HCAL_ENDCAP;
+    else
+      pfrh.layer() = PFLayer::NONE;
+  }
+
+  template <>
+  ALPAKA_FN_ACC void PFRecHitProducerKernelConstruct<ECAL>::constructPFRecHit(
+      reco::PFRecHitDeviceCollection::View::element pfrh,
+      const ECAL::CaloRecHitSoATypeDevice::ConstView::const_element rh) {
+    pfrh.detId() = rh.detId();
+    pfrh.energy() = rh.energy();
+    pfrh.time() = rh.time();
+    pfrh.depth() = 1;
+    const uint32_t subdet = getSubdet(pfrh.detId());
+    if (subdet == EcalBarrel)
+      pfrh.layer() = PFLayer::ECAL_BARREL;
+    else if (subdet == EcalEndcap)
+      pfrh.layer() = PFLayer::ECAL_ENDCAP;
+    else
+      pfrh.layer() = PFLayer::NONE;
+  }
+
+  // Kernel to associate topology information of PFRecHits
+  template <typename CAL>
+  struct PFRecHitProducerKernelTopology {
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC void operator()(const TAcc& acc,
+                                  const typename CAL::TopologyTypeDevice::ConstView topology,
+                                  reco::PFRecHitDeviceCollection::View pfRecHits,
+                                  const uint32_t* __restrict__ denseId2pfRecHit,
+                                  uint32_t* __restrict__ num_pfRecHits) const {
+      // First thread updates size field pfRecHits SoA
+      if (const int32_t thread = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0u]; thread == 0)
+        pfRecHits.size() = *num_pfRecHits;
+
+      // Assign position information and associate neighbours
+      for (int32_t i : cms::alpakatools::elements_with_stride(acc, *num_pfRecHits)) {
+        const uint32_t denseId = CAL::detId2denseId(pfRecHits.detId(i));
+
+        pfRecHits.x(i) = topology.positionX(denseId);
+        pfRecHits.y(i) = topology.positionY(denseId);
+        pfRecHits.z(i) = topology.positionZ(denseId);
+
+        for (uint32_t n = 0; n < 8; n++) {
+          pfRecHits.neighbours(i)(n) = -1;
+          const uint32_t denseId_neighbour = topology.neighbours(denseId)(n);
+          if (denseId_neighbour != 0xffffffff) {
+            const uint32_t pfRecHit_neighbour = denseId2pfRecHit[denseId_neighbour];
+            if (pfRecHit_neighbour != 0xffffffff)
+              pfRecHits.neighbours(i)(n) = (int32_t)pfRecHit_neighbour;
+          }
+        }
```

Comment:
For the moment this is just an idea, not a suggestion to implement any changes.

How much time is spent in this kernel ?
If it's negligible, we don't need to change anything.
If it's significant, we could look into a couple of options:
  - inverting the loop, with an outer loop over `n` and an inner loop over `i`;
  - parallelising this loop using a 2D grid.

Reply:
This could probably be optimised. But the entire module is already extremely fast (~10us/event, maybe 20% of that in this loop), so I think this is probably not worth worrying about.